File: backend/app/swarm/providers.py
"""Model provider factory: Bedrock, Anthropic, OpenAI-compatible, vLLM, LiteLLM, Ollama, Demo."""

from typing import Any


def build_model(settings: dict, override: dict | None = None):
    cfg = dict(settings)
    if override:
        cfg.update({k: v for k, v in override.items() if v})
    provider = (cfg.get("provider") or "bedrock").lower()
    model_id = cfg.get("model_id") or ""
    temperature = cfg.get("temperature")
    max_tokens = int(cfg.get("max_tokens") or 4096)
    api_key = cfg.get("api_key") or ""
    base_url = (cfg.get("base_url") or "").rstrip("/")

    if provider == "bedrock":
        from strands.models import BedrockModel

        kwargs: dict[str, Any] = {"model_id": model_id or "global.anthropic.claude-sonnet-5",
                                  "region_name": cfg.get("region") or "us-east-1", "max_tokens": max_tokens}
        if temperature is not None:
            kwargs["temperature"] = float(temperature)
        if cfg.get("guardrail_id"):
            kwargs.update(guardrail_id=cfg["guardrail_id"], guardrail_version=cfg.get("guardrail_version") or "DRAFT",
                          guardrail_trace="enabled")
        return BedrockModel(**kwargs)

    params: dict[str, Any] = {"max_tokens": max_tokens}
    if temperature is not None:
        params["temperature"] = float(temperature)

    if provider == "anthropic":
        from strands.models.anthropic import AnthropicModel

        return AnthropicModel(client_args={"api_key": api_key}, model_id=model_id or "claude-sonnet-5",
                              max_tokens=max_tokens, params={k: v for k, v in params.items() if k != "max_tokens"})

    if provider in {"openai", "vllm", "nvidia_nim"}:
        from strands.models.openai import OpenAIModel

        client_args: dict[str, Any] = {"api_key": api_key or "EMPTY"}
        if base_url:
            client_args["base_url"] = base_url
        elif provider == "vllm":
            client_args["base_url"] = "http://vllm:8000/v1"
        elif provider == "nvidia_nim":
            client_args["base_url"] = "https://integrate.api.nvidia.com/v1"
        return OpenAIModel(client_args=client_args,
                           model_id=model_id or {"vllm": "Qwen/Qwen3-32B", "nvidia_nim": "nvidia/nemotron-3-super-120b-a12b"}.get(
                               provider, "gpt-5.4"),
                           params=params)

    if provider == "litellm":
        from strands.models.litellm import LiteLLMModel

        client_args = {}
        if api_key:
            client_args["api_key"] = api_key
        if base_url:
            client_args["api_base"] = base_url
        return LiteLLMModel(client_args=client_args, model_id=model_id or "openai/gpt-5.4", params=params)

    if provider == "ollama":
        from strands.models.ollama import OllamaModel

        # Ollama defaults to a 2048-token context, far too small for multi-step tool calling.
        kwargs = {"model_id": model_id or "qwen3:32b", "max_tokens": max_tokens,
                  "options": {"num_ctx": int(cfg.get("context_window") or 32768)}}
        if temperature is not None:
            kwargs["temperature"] = float(temperature)
        return OllamaModel(base_url or "http://ollama:11434", **kwargs)

    if provider == "demo":
        import os

        from .demo_model import DemoModel

        lo, _, hi = os.getenv("AUTOPILOT_DEMO_LATENCY_MS", "350,1100").partition(",")
        return DemoModel(model_id=model_id or "autopilot-demo-1",
                         latency=(int(lo) / 1000, int(hi or lo) / 1000))

    raise ValueError(f"Unknown model provider '{provider}'")

File: ui/src/lib/providers.ts
import type { ModelSettings, ProviderInfo } from "@/api/types";

/** Fallback used when /settings/providers isn't reachable (e.g. during first-run setup). */
export const FALLBACK_PROVIDERS: ProviderInfo[] = [
  {
    id: "bedrock",
    label: "Amazon Bedrock",
    description: "Default for enterprise. IAM credentials, in-account inference, optional Bedrock Guardrails.",
    fields: ["model_id", "region", "guardrail_id", "guardrail_version"],
    defaults: { model_id: "global.anthropic.claude-sonnet-5", region: "us-east-1" },
    local: false,
  },
  {
    id: "anthropic",
    label: "Anthropic",
    description: "Claude models via the Anthropic API.",
    fields: ["model_id", "api_key"],
    defaults: { model_id: "claude-sonnet-5" },
    local: false,
  },
  {
    id: "openai",
    label: "OpenAI-compatible",
    description: "OpenAI, Azure OpenAI, NVIDIA NIM or any OpenAI-compatible endpoint.",
    fields: ["base_url", "model_id", "api_key"],
    defaults: { base_url: "https://api.openai.com/v1", model_id: "gpt-5.4" },
    local: false,
  },
  {
    id: "nvidia_nim",
    label: "NVIDIA NIM",
    description: "NVIDIA Nemotron and other NIM models via build.nvidia.com, or a self-hosted NIM.",
    fields: ["base_url", "model_id", "api_key"],
    defaults: { base_url: "https://integrate.api.nvidia.com/v1", model_id: "nvidia/nemotron-3-super-120b-a12b" },
    local: false,
  },
  {
    id: "vllm",
    label: "vLLM",
    description: "Self-hosted GPU inference through vLLM's OpenAI-compatible server.",
    fields: ["base_url", "model_id", "api_key"],
    defaults: { base_url: "http://vllm:8000/v1", model_id: "Qwen/Qwen3-32B" },
    local: true,
  },
  {
    id: "litellm",
    label: "LiteLLM",
    description: "Route through a LiteLLM proxy, or call 100+ providers directly.",
    fields: ["base_url", "model_id", "api_key"],
    defaults: { base_url: "http://litellm:4000", model_id: "openai/gpt-5.4" },
    local: false,
  },
  {
    id: "ollama",
    label: "Ollama",
    description: "Air-gapped local models. Nothing leaves the host.",
    fields: ["base_url", "model_id", "context_window"],
    defaults: { base_url: "http://ollama:11434", model_id: "qwen3:32b", context_window: 32768 },
    local: true,
  },
  {
    id: "demo",
    label: "Demo (scripted)",
    description: "Deterministic scripted model for demos and CI. No API key, no network.",
    fields: [],
    defaults: { model_id: "demo-soc-1" },
    local: true,
  },
];

export const FIELD_LABELS: Partial<Record<keyof ModelSettings, { label: string; placeholder?: string; secret?: boolean }>> = {
  model_id: { label: "Model ID" },
  base_url: { label: "Base URL", placeholder: "http://host:port/v1" },
  api_key: { label: "API key", secret: true },
  region: { label: "AWS region", placeholder: "us-east-1" },
  guardrail_id: { label: "Bedrock Guardrail ID (optional)" },
  guardrail_version: { label: "Guardrail version", placeholder: "DRAFT" },
  temperature: { label: "Temperature" },
  max_tokens: { label: "Max tokens" },
  context_window: { label: "Context window (tokens)", placeholder: "32768" },
};

export const AUTONOMY_LEVELS = [
  {
    id: "observe" as const,
    label: "Observe",
    tag: "L0",
    desc: "Agents triage and investigate only. No response actions are proposed.",
  },
  {
    id: "recommend" as const,
    label: "Recommend",
    tag: "L1",
    desc: "Agents propose actions. A human approves and a human executes (two-tier).",
  },
  {
    id: "supervised" as const,
    label: "Supervised",
    tag: "L2",
    desc: "A human approves. The platform executes and verifies automatically.",
  },
  {
    id: "autonomous" as const,
    label: "Autonomous",
    tag: "L3",
    desc: "Allow-listed, high-confidence actions on non-protected targets run automatically. Everything else is supervised.",
  },
];
